File: cs/Minsk.Tests/CodeAnalysis/Syntax/LexerTests.cs
using Minsk.CodeAnalysis.Syntax;
using Xunit;

namespace Minsk.Tests.CodeAnalysis.Syntax;

public sealed class LexerTests
{
    private static IEnumerable<SimpleToken> GetTokens()
    {
        var fixedTokens = Enum.GetValues(typeof(SyntaxKind))
            .Cast<SyntaxKind>()
            .Select(k => (kind: k, text: SyntaxFacts.GetText(k)))
            .Where(t => t.text != null)
            .Select(t => new SimpleToken(t.kind, t.text!));

        var dynamicTokens = new SimpleToken[]
        {
            new(SyntaxKind.IdentifierToken, "a"),
            new(SyntaxKind.IdentifierToken, "abc"),
            new(SyntaxKind.NumberToken, "1"),
            new(SyntaxKind.NumberToken, "123")
        };

        return fixedTokens.Concat(dynamicTokens);
    }

    private static IEnumerable<SimpleToken> GetSeparators()
    {
        return new SimpleToken[]
        {
            new(SyntaxKind.WhitespaceToken, " "),
            new(SyntaxKind.WhitespaceToken, "  "),
            new(SyntaxKind.WhitespaceToken, "\n"),
            new(SyntaxKind.WhitespaceToken, "\r"),
            new(SyntaxKind.WhitespaceToken, "\r\n")
        };
    }

    private static IEnumerable<(SimpleToken t1, SimpleToken t2)> GetTokenPairs()
    {
        return GetTokens()
            .SelectMany(_ => GetTokens(), (t1, t2) => new { t1, t2 })
            .Where(elem => !RequiresSeparator(elem.t1.Kind, elem.t2.Kind))
            .Select(elem => (elem.t1, elem.t2));
    }

    private static IEnumerable<(SimpleToken t1, SimpleToken separator, SimpleToken t2)> GetTokenPairsWithSeparator()
    {
        return GetTokens()
            .SelectMany(_ => GetTokens(), (t1, t2) => new { t1, t2 })
            .Where(elem => RequiresSeparator(elem.t1.Kind, elem.t2.Kind))
            .SelectMany(_ => GetSeparators(), (elem, separator) => (elem.t1, separator, elem.t2));
    }

    private static bool RequiresSeparator(SyntaxKind t1Kind, SyntaxKind t2Kind)
    {
        var t1IsKeyword = t1Kind.ToString().EndsWith("Keyword");
        var t2IsKeyword = t2Kind.ToString().EndsWith("Keyword");

        if ((t1Kind == SyntaxKind.IdentifierToken || t1IsKeyword) &&
            (t2Kind == SyntaxKind.IdentifierToken || t2IsKeyword))
        {
            return true;
        }

        if ((t1Kind is SyntaxKind.NumberToken or SyntaxKind.IdentifierToken || t1IsKeyword) &&
            t2Kind == SyntaxKind.NumberToken)
        {
            return true;
        }

        if (t1Kind is SyntaxKind.BangToken or SyntaxKind.EqualsToken &&
            t2Kind is SyntaxKind.EqualsToken or SyntaxKind.EqualsEqualsToken)
        {
            return true;
        }

        return false;
    }

    private static IEnumerable<object[]> GetLexesTokenData()
    {
        return GetTokens().Select(t => new object[] { t });
    }

    private static IEnumerable<object[]> GetLexesTokenPairsData()
    {
        return GetTokenPairs().Select(t => new object[] { t.t1, t.t2 });
    }

    private static IEnumerable<object[]> GetLexesTokenPairsWithSeparatorData()
    {
        return GetTokenPairsWithSeparator().Select(t => new object[] { t.t1, t.separator, t.t2 });
    }

    [Fact]
    private void TestsAllTokens()
    {
        var tokenKinds = Enum.GetValues(typeof(SyntaxKind))
            .Cast<SyntaxKind>()
            .Where(k => k.ToString().EndsWith("Keyword") || k.ToString().EndsWith("Token"));

        var testedTokenKinds = GetTokens().Concat(GetSeparators()).Select(t => t.Kind);

        var untestedTokenKinds = new SortedSet<SyntaxKind>(tokenKinds);
        untestedTokenKinds.Remove(SyntaxKind.BadToken);
        untestedTokenKinds.Remove(SyntaxKind.EndOfFileToken);
        untestedTokenKinds.ExceptWith(testedTokenKinds);

        Assert.Empty(untestedTokenKinds);
    }

    [Theory]
    [MemberData(nameof(GetLexesTokenData))]
    private void LexesToken(SimpleToken t)
    {
        var tokens = SyntaxTree.ParseTokens(t.Text);
        var token = Assert.Single(tokens);
        Assert.Equal(t.Kind, token.Kind);
        Assert.Equal(t.Text, token.Text);
    }

    [Theory]
    [MemberData(nameof(GetLexesTokenPairsData))]
    private void LexesTokenPairs(SimpleToken t1, SimpleToken t2)
    {
        var text = t1.Text + t2.Text;
        var tokens = SyntaxTree.ParseTokens(text).ToArray();
        Assert.Equal(2, tokens.Length);
        Assert.Equal(t1.Kind, tokens[0].Kind);
        Assert.Equal(t1.Text, tokens[0].Text);
        Assert.Equal(t2.Kind, tokens[1].Kind);
        Assert.Equal(t2.Text, tokens[1].Text);
    }

    [Theory]
    [MemberData(nameof(GetLexesTokenPairsWithSeparatorData))]
    private void LexesTokenPairsWithSeparator(SimpleToken t1, SimpleToken separator, SimpleToken t2)
    {
        var text = t1.Text + separator.Text + t2.Text;
        var tokens = SyntaxTree.ParseTokens(text).ToArray();
        Assert.Equal(3, tokens.Length);
        Assert.Equal(t1.Kind, tokens[0].Kind);
        Assert.Equal(t1.Text, tokens[0].Text);
        Assert.Equal(separator.Kind, tokens[1].Kind);
        Assert.Equal(separator.Text, tokens[1].Text);
        Assert.Equal(t2.Kind, tokens[2].Kind);
        Assert.Equal(t2.Text, tokens[2].Text);
    }

    private sealed class SimpleToken
    {
        public SimpleToken(SyntaxKind kind, string text)
        {
            Kind = kind;
            Text = text;
        }

        public SyntaxKind Kind { get; }
        public string Text { get; }
    }
}

File: kt/Minsk/minsk/src/test/kotlin/dev/phytolizer/minsk/analysis/syntax/LexerTests.kt
package dev.phytolizer.minsk.analysis.syntax

import dev.phytolizer.minsk.TestUtils
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe

class LexerTests : FunSpec({
    context("lexes token") {
        withData(tokens()) { (kind, text) ->
            val tokens = SyntaxTree.parseTokens(text)
            tokens.size shouldBe 1
            tokens[0].kind shouldBe kind
            tokens[0].text shouldBe text
        }
    }

    context("lexes token pairs") {
        withData(
            nameFn = { "${it[0].kind} (${it[0].text}) x ${it[1].kind} (${it[1].text})" },
            tokenPairs(),
        ) { tt ->
            val text = "${tt[0].text}${tt[1].text}"
            val tokens = SyntaxTree.parseTokens(text)
            tokens.size shouldBe 2
            tokens[0].kind shouldBe tt[0].kind
            tokens[0].text shouldBe tt[0].text
            tokens[1].kind shouldBe tt[1].kind
            tokens[1].text shouldBe tt[1].text
        }
    }

    context("lexes token pairs with separator") {
        withData(
            nameFn = { "${it[0].kind} (${it[0].text}) x sep x ${it[2].kind} (${it[2].text})" },
            tokenPairsWithSeparator(),
        ) { tt ->
            val text = "${tt[0].text}${tt[1].text}${tt[2].text}"
            val tokens = SyntaxTree.parseTokens(text)
            tokens.size shouldBe 3
            tokens[0].kind shouldBe tt[0].kind
            tokens[0].text shouldBe tt[0].text
            tokens[1].kind shouldBe tt[1].kind
            tokens[1].text shouldBe tt[1].text
            tokens[2].kind shouldBe tt[2].kind
            tokens[2].text shouldBe tt[2].text
        }
    }
}) {
    companion object {
        private fun tokens(): List<SimpleToken> {
            val fixedTokens = SyntaxKind.values().filter { SyntaxFacts.getText(it) != null }
                .map { SimpleToken(it, SyntaxFacts.getText(it)!!) }
            return listOf(
                fixedTokens,
                listOf(
                    SimpleToken(SyntaxKind.IdentifierToken, "a"),
                    SimpleToken(SyntaxKind.IdentifierToken, "abc"),
                    SimpleToken(SyntaxKind.NumberToken, "1"),
                    SimpleToken(SyntaxKind.NumberToken, "123"),
                ),
            ).flatten()
        }

        private fun separators(): List<SimpleToken> = listOf(
            SimpleToken(SyntaxKind.WhitespaceToken, " "),
            SimpleToken(SyntaxKind.WhitespaceToken, "  "),
            SimpleToken(SyntaxKind.WhitespaceToken, "\r"),
            SimpleToken(SyntaxKind.WhitespaceToken, "\n"),
            SimpleToken(SyntaxKind.WhitespaceToken, "\r\n"),
        )

        private fun tokenPairs() =
            TestUtils.cartesianProduct(tokens(), tokens()).filter { !requiresSeparator(it[0], it[1]) }

        private fun tokenPairsWithSeparator() = TestUtils.cartesianProduct(
            tokens(),
            separators(),
            tokens(),
        ).filter { requiresSeparator(it[0], it[2]) }

        private fun requiresSeparator(t1: SimpleToken, t2: SimpleToken): Boolean {
            val t1IsKeyword = t1.kind.toString().endsWith("Keyword")
            val t2IsKeyword = t2.kind.toString().endsWith("Keyword")

            return if ((t1.kind == SyntaxKind.IdentifierToken || t1IsKeyword) && (t2.kind == SyntaxKind.IdentifierToken || t2IsKeyword)) {
                true
            } else if ((t1.kind == SyntaxKind.NumberToken || t1.kind == SyntaxKind.IdentifierToken || t1IsKeyword) && t2.kind == SyntaxKind.NumberToken) {
                true
            } else (t1.kind == SyntaxKind.BangToken || t1.kind == SyntaxKind.EqualsToken) && (t2.kind == SyntaxKind.EqualsToken || t2.kind == SyntaxKind.EqualsEqualsToken)
        }
    }

    data class SimpleToken(val kind: SyntaxKind, val text: String)
}

